crates/evm/src/precompile_cache.rs

```diff
@@ -0,0 +1,361 @@
+    #[cfg(feature = "metrics")]
+    metrics: CachedPrecompileMetrics,
+}
+
+impl<P> MaybeCachedPrecompileProvider<P> {
+    /// Given a [`PrecompileProvider`]  and cache for a specific precompile provider,
+    /// create a cached wrapper that can be used inside Evm.
+    #[cfg(feature = "std")]
+    pub fn new_with_cache(
+        precompile_provider: P,
+        cache: alloc::sync::Arc<PrecompileCache>,
+    ) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: Some(cache),
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+
+    /// Creates a new `MaybeCachedPrecompileProvider` with cache disabled.
+    pub fn new_without_cache(precompile_provider: P) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: None,
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+}
+
+impl<CTX: ContextTr, P: PrecompileProvider<CTX, Output = InterpreterResult>> PrecompileProvider<CTX>
+    for MaybeCachedPrecompileProvider<P>
+{
+    type Output = P::Output;
+
+    fn set_spec(&mut self, spec: <CTX::Cfg as Cfg>::Spec) -> bool {
+        let old_spec = self.spec;
+        self.spec = spec.clone().into();
+
+        if self.precompile_provider.set_spec(spec) {
+            return true;
+        }
+
+        old_spec != self.spec
+    }
+
+    #[cfg(feature = "std")]
+    fn run(
+        &mut self,
+        context: &mut CTX,
+        address: &Address,
+        inputs: &InputsImpl,
+        is_static: bool,
+        gas_limit: u64,
+    ) -> Result<Option<Self::Output>, String> {
+        use revm::interpreter::{Gas, InstructionResult};
+
+        if let Some(cache) = &self.cache {
+            let key =
+                PrecompileKey { address: *address, spec: self.spec, input: inputs.input.clone() };
+
+            let cache_result = cache.cache.get(&key);
+
+            if let Some(ref entry) = cache_result {
+                // if gas_limit is below known lower bound, we know it will fail with OOG
+                if gas_limit <= entry.lower_gas_limit {
+                    self.increment_by_one_precompile_cache_hits();
+
+                    let result = InterpreterResult {
+                        result: InstructionResult::PrecompileOOG,
+                        gas: Gas::new(gas_limit),
+                        output: alloy_primitives::Bytes::new(),
+                    };
+
+                    return Ok(Some(result));
+                }
+
+                // if gas_limit is above upper bound, use the cached result
+                if gas_limit >= entry.upper_gas_limit {
+                    self.increment_by_one_precompile_cache_hits();
+
+                    // for successful results, we need to ensure gas costs are correct when
```

Comment:
why only for successfuly result?

Reply:
good point, it is the only case in which we are recording gas cost https://github.com/bluealloy/revm/blob/35a0611281997608dc9d1bd655b348105334f451/crates/handler/src/precompile_provider.rs#L111

Reply:
added comment about this